src/ophys_etl/modules/mesoscope_splitting/tiff_splitter.py

```diff
@@ -379,77 +467,117 @@ class TimeSeriesSplitter(ScanImageTiffSplitter):
-        msg += "data directly to an HDF5 file."
-        raise NotImplementedError(msg)
-
-    def write_output_file(self,
-                          i_roi: int,
-                          z_value: float,
-                          output_path: pathlib.Path) -> None:
+    def write_output_files(self,
+                           output_path_map: Dict[Tuple[int, float],
+                                                 pathlib.Path],
+                           tmp_dir: Optional[pathlib.Path] = None,
+                           dump_every: int = 1000,
+                           logger: Optional[callable] = None) -> None:
         """
         Write all of the pages associated with an
         (i_roi, z_value) pair to an HDF5 file.
 
         Parameters
         ----------
-        i_roi: int
-            index of the ROI
+        output_path_map: Dict[Tuple[int, float], pathlib.Path]
+            Dict mapping (i_roi, z_value) pairs to output paths
+            where the data for those ROIs should go.
 
-        z_value: int
-            z value of the scanfield plane
+        tmp_dir: Optional[pathlib.Path]
+            Directory where temporary files will be written.
 
-        output_path: pathlib.Path
-            path to the HDF5 file to be written
+        dump_every: int
+            Number of frames to store in each temprorary file
+            (see Notes)
+
+        logger: Optional[callable]
+            Logger which will be invoked with logger.INFO
+            by worker methods
 
         Returns
         -------
         None
-            output is written to output_path
+            Timeseries for the ROIs are written to the paths
+            specified in output_path_map.
+
+            If not specified, will write temporary files into
+            the directory where the final files are meant to
+            be written.
+
+        Notes
+        -----
+        Because the only way to get n_pages from a BigTIFF is to
+        iterate over its pages, counting, this module works by iterating
+        over the pages, splitting the timeseries data into small temp files
+        as it goes and keeping track of how many total pages are being written
+        for each ROI. After the temp files are written, the temp files are
+        gathered together into the final files specified in output_path_map
+        and the temporary files are deleted.
         """
 
-        if output_path.suffix != '.h5':
-            msg = "expected HDF5 output path; "
-            msg += f"you gave {output_path.resolve().absolute()}"
+        for key_pair in output_path_map:
+            output_path = output_path_map[key_pair]
+            if output_path.suffix != '.h5':
+                msg = "expected HDF5 output path; "
+                msg += f"you gave {output_path.resolve().absolute()}"
+                raise ValueError(msg)
+
+            i_roi = key_pair[0]
+            z_value = key_pair[1]
+
+            if i_roi >= self.n_rois:
```

Comment:
Any case where this could be less than?

Reply:
Sorry: I'm not sure what the ask is here.

What I'm checking here is that i_roi is a valid value (i.e. that it is between 0 and self.n_rois-1). I am emitting a custom error message so that whoever debugs a failure gets a little more info than the default python IndexError.

Reply:
Okay, go ahead and keep it. I guess what I was pointing out as that `-1` in your test is considered a valid `i_roi` value.

Reply:
Oh. I see. That makes sense. I'll add -1 as an invalid value

Reply:
Right, anything negative would be valid with the current check.